timeline-tests/timeline-tests.cabal

```diff
@@ -0,0 +1,38 @@
+cabal-version: 2.2
+name:          timeline-tests
```

Comment:
Why is this a separate library?

Reply:
Can it be put into `timeline`, while depending on `timeline-hedgehog`? I thought this would result in cyclic dependency.

Reply:
You _might_ be okay, because there is no dependency cycle at the component level. Tooling that does package-level dependency resolution (example: `nixpkgs`; anti-example: `haskell.nix`) will need to break the loop with something like `dontCheck`. Worth checking.

Reply:
Haskell infrastructure in nixpkgs can't handle this situation.
```
error: infinite recursion encountered
```

Now we have two options:
1. Use `dontCheck`.
    1. We can't build and test all in a simple command like `nix build`. We need to enter the dev shell and run cabal tests. This complicates our CI setup.
    2. Tests won't be run automatically for our users like a normal hackage package. 
2. Keep tests in `timeline-tests` package
    1. `nix build` builds and tests everything for us.
    2. Tests won't be run automatically for our users like a normal hackage package. This is the same as using `dontCheck`.

As a result, `timeline-tests` is a win, unless we have a better approach. (Haskell really need to learn from Rust's toolchain: cargo ["features"](https://doc.rust-lang.org/cargo/reference/features.html) solves this very well)

Reply:
Having a third package that people need to remember to run, just to test anything at all, seems bad. I therefore still think `dontCheck` is better, because it's nixpkgs' problem that it's not handling such setups properly. Also, `timeline-tests` is not a package that wants to go on Hackage, and it's pretty messy.

You could also split the tests: put any unit tests into `timeline` using `tasty`/`tasty-hunit`, and put property tests into `timeline-hedgehog` (since they're kinda testing the generator as well as `timeline`).

Reply:
Oh no I didn't test cabal... and [cabal itself doesn't work, either](https://github.com/haskell/cabal/issues/4087).

But I got your concern. I'll try other approaches.

Reply:
Split to `timeline`, `timeline-core` and `timeline-hedgehog`. `timeline-hedgehog` only depends on `timeline-core`, and test suite in `timeline` depends on `timeline-hedgehog`. This eliminate the package level cyclic dependency.

`timeline-core` only exposes `Data.Timeline.Internal`, which tells the users that it should not be used directly.

Reply:
The same argument against `hedgehog` being a dependency of `timeline` and therefore "we should split packages" is equally applicable to `tasty-golden`. I'm just not convinced that splitting this into three packages really gives you that much, when something like `retry` already pulls hedgehog into a build graph.

I don't see `timeline` -> `timeline-hedgehog` -> `timeline-core` bringing that much benefit over `timeline`, when it imposes a bunch of ongoing maintenance nuisance keeping bounds updated and so forth. I could be persuaded that `hedgehog` is really dependency-hungry, in which case `timeline-hedgehog` (generators and property tests) -> `timeline` (core and unit tests) seems like a good compromise.

Also, have you seen https://hackage.haskell.org/package/int-interval-map ? I stumbled upon it looking at hedgehog's reverse dependencies.

Reply:
When a user depend on `timeline` and doesn't enable the build of tests in dependencies, `timeline-hedgehog` is not built, it's just the dependency of the test-suite section.

What we need to choose is, should we force `hedgehog` on users (and possibly `QuickCheck` in the future) even if they only use the library component? The price to not forcing `hedgehog` is uploading 3 packages to Hackage every time, but that could be solved by a script.

Reply:
> Also, have you seen https://hackage.haskell.org/package/int-interval-map ? I stumbled upon it looking at hedgehog's reverse dependencies.

It's not directly usable. But looks interesting, maybe we can extend on it.

Reply:
> What we need to choose is, should we force `hedgehog` on users (and possibly `QuickCheck` in the future) even if they only use the library component? The price to not forcing `hedgehog` is uploading 3 packages to Hackage every time, but that could be solved by a script.

Why not put the property tests in `timeline-hedgehog` then? It separates things without introducing too many additional packages, and CI can run the tests regardless of the package they belong to.

`QuickCheck` is harder because a) it forgoes automatic shrinking, and b) you have to choose between a direct `QuickCheck` dependency or orphan instances.

Reply:
TL;DR: given all factors we know for now, I'll merge them into one package.

> Why not put the property tests in timeline-hedgehog then? It separates things without introducing too many additional packages, and CI can run the tests regardless of the package they belong to.

This seems to sit in somewhere between putting all tests in timeline, and all tests in timeline-tests. In the first case, users depending only on timeline will be able to run all tests on timeline. In the second case, the tests are only run in our CI. The question is, Why do we want the users to run our tests? If this is viewed as a safeguard to make sure our code behaves correctly, then migrating tests from timeline  to timeline-hedgehog will weaken this safeguard. Also, tests for a single function are split into different packages, which makes writing tests harder because we have to look at both places.

There seems to be no perfect way to make parts of a library optional in Haskell, which is a pity. I tried to implement the most general approach. But yeah, given that no one else is using or complaining the `hedgehog` dependency, we should probably go with a single package and consider more robust way in the future.

Reply:
And https://github.com/haskell/cabal/issues/5660 might be the ultimate solution. I should see it that can be pushed forward.